### PR TITLE
feat: show tax information on public ticket page

### LINKED ISSUE #2082: No Tax Information being shown on public tickets
**Describe the bug**

Even after setting up tax information for an event, we can't see it being shown up on the event's ticket page or anywhere later in ticket process.

**To Reproduce**
1. Set taxes in an event
2. Go to event page
3. Scroll down to Tickets

**Expected behavior**

A separate Tax Information must be shown which shows the tax rate and related information

**Screenshots**

taxinfo
ticketwithouttax

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Ubuntu 16.04 LTS
 - Browser [e.g. chrome, safari] Chrome
 
**Additional context**

Can be worked upon by anyone

### PR DESCRIPTION
**Pull request by uds5501**
Fixes #2082 

#### Short description of what this resolves:
Implements the tax information details show in the public ticket page just like eventbrite

#### Changes proposed in this pull request:

- add design changes in accordance with eventbrite (see issue for design similarity)
- add `taxIncurred` attribute in ticket object

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### Screenshots : 
Event with tax:
image

Discount code activate:
image

Event with tax included in price
image

### CODE REVIEW

**Comment by shreyanshdwivedi** [comment]
@uds5501 can you please add screenshots?

**Comment by uds5501** [comment]
@shreyanshdwivedi was adding them, wait xD

**Comment by uds5501** [comment]
event with tax :
image

@shreyanshdwivedi @mrsaicharan1 @kushthedude Peer reviews please!

File: app/templates/components/public/ticket-list.hbs [hunk 1]
```diff
@@ -29,10 +29,28 @@
...
+                  {{/if}}
                 </div>
               </td>
             {{else}}
-              <td id="{{ticket.id}}_price">{{currency-symbol eventCurrency}} {{format-number ticket.price}}</td>
+              <td id="{{ticket.id}}_price">
+                {{currency-symbol eventCurrency}} {{format-number ticket.price}}
+                {{#if taxInfo}}
+                  <small class="ui gray-text small">
+                    + {{currency-symbol eventCurrency}} {{format-number (sub ticket.taxIncurred ticket.price)}}
```
Comment (shreyanshdwivedi) [inline app/templates/components/public/ticket-list.hbs]: @uds5501 I think you should add a condition here to remove displaying something like `US$ 0` tax for free tickets

Comment (shreyanshdwivedi) [reply app/templates/components/public/ticket-list.hbs]: And can you add a screenshot of situation if a discount code is applied

Comment (uds5501) [reply app/templates/components/public/ticket-list.hbs]: @shreyanshdwivedi with the discount:

image
and I will take care of the free tickets in next commit

Comment (shreyanshdwivedi) [reply app/templates/components/public/ticket-list.hbs]: @uds5501 is your tax percent or amount? Of percent then it should be applied to the discounted amount. Right?

Comment (uds5501) [reply app/templates/components/public/ticket-list.hbs]: @shreyanshdwivedi 
Tax is always percent rate. 
- The way i implemented it right now is calculate tax on base price of the ticket (which doesn't change)
- If a discount code is applied, it deducts the total price.

Comment (shreyanshdwivedi) [reply app/templates/components/public/ticket-list.hbs]: @uds5501 if it is percent, we should ensure to apply tax on what user pays. This is basically followed everywhere.

Comment (uds5501) [reply app/templates/components/public/ticket-list.hbs]: @shreyanshdwivedi okay, wip-ing it till i get this right.

File: app/controllers/public/index.js [hunk 2]
```diff
@@ -113,13 +113,6 @@ export default Controller.extend({
         return;
       }
       let { order, event } = this.model;
-      let tax =  await event.get('tax');
-      if (tax) {
```
Comment (kushthedude) [inline app/controllers/public/index.js]: @uds5501 Why did you remove the following ?

File: app/controllers/public/index.js [hunk 3]
```diff
@@ -113,13 +113,6 @@ export default Controller.extend({
...
-        if (!tax.get('isTaxIncludedInPrice')) {
```
Comment (kushthedude) [inline app/controllers/public/index.js]: @uds5501 Why did you remove the following ?

File: app/templates/components/public/ticket-list.hbs (hunk 1 above)
Comment (kushthedude) [reply app/templates/components/public/ticket-list.hbs]: @uds5501 What will be the message in case tax is included in price and what about if we have added some ticket fees ?

File: app/controllers/public/index.js (hunk 3 above)
Comment (uds5501) [reply app/controllers/public/index.js]: @kushthedude because it was adding tax on tax. I will take care of this

Comment (uds5501) [reply app/controllers/public/index.js]: @kushthedude accidentally resolved the other one. The ticket fees is something which organisers have to handle, not the ticket buyers. It's  more like a cut for the platform (@CosmicCoder96 please correct me if I am wrong)

Comment (kushthedude) [reply app/controllers/public/index.js]: > @kushthedude because it was adding tax on tax. I will take care of this

Removing this will break the order summary page . As it was responsible for tax modal display in order wizard

Comment (uds5501) [reply app/controllers/public/index.js]: @kushthedude it won't. I double checked that. you can go ahead and try it in this PRs deployment itself.

**Comment by uds5501** [comment]
@shreyanshdwivedi @kushthedude It's ready to review now. Please have a look.

File: app/components/public/ticket-list.js [hunk 4]
```diff
@@ -93,12 +105,22 @@ export default Component.extend(FormMixin, {
           this.order.set('discountCode', discountCode);
           let tickets = await discountCode.get('tickets');
           tickets.forEach(ticket => {
+            console.log(ticket);
```
Comment (kushthedude) [inline app/components/public/ticket-list.js]: @uds5501 Why the following

File: app/templates/public/index.hbs [hunk 5]
```diff
@@ -20,6 +20,7 @@
                                  order=model.order
                                  attendees=model.attendees
                                  code=code
+                                 taxInfo=model.event.tax
```
Comment (kushthedude) [inline app/templates/public/index.hbs]: @uds5501 Please modify the test of `Public/Ticket List` to include the `Tax Object`

File: app/components/public/ticket-list.js [hunk 6]
```diff
@@ -29,14 +29,20 @@ export default Component.extend(FormMixin, {
   }),
   hasTicketsInOrder: computed('tickets.@each.orderQuantity', function() {
     return sumBy(this.tickets.toArray(),
-      ticket => ticket.getWithDefault('orderQuantity', 0)
+      ticket => (ticket.orderQuantity || 0)
```
Comment (CosmicCoder96) [inline app/components/public/ticket-list.js]: Why?

File: app/components/public/ticket-list.js [hunk 7]
```diff
@@ -29,14 +29,20 @@ export default Component.extend(FormMixin, {
...
     ) > 0;
   }),
 
   total: computed('tickets.@each.orderQuantity', 'tickets.@each.discount', function() {
-    return sumBy(this.tickets.toArray(),
-      ticket => (ticket.getWithDefault('price', 0) - ticket.getWithDefault('discount', 0)) * ticket.getWithDefault('orderQuantity', 0)
-    );
```
Comment (CosmicCoder96) [inline app/components/public/ticket-list.js]: Why are you replacing getWithDefault?

File: app/controllers/public/index.js (hunk 3 above)
Comment (pradeepgangwar) [reply app/controllers/public/index.js]: @uds5501 Is it sending the amount back to server? Because last time it was added here was because it was not reflecting the taxed amount on server and everything was correct on frontend Please check it again that after order too correct amount is getting saved both on server and FE.

File: app/models/ticket.js [hunk 8]
```diff
@@ -44,5 +44,13 @@ export default ModelBase.extend({
 
   itemTotal: computed('price', 'quantity', function() {
     return this.price * this.quantity;
+  }),
+
+  taxIncurred: computed('event.tax', function() {
+    let taxType = this.event.get('tax.isTaxIncludedInPrice');
```
Comment (pradeepgangwar) [inline app/models/ticket.js]: Use ES6 getters.

File: app/models/ticket.js [hunk 9]
```diff
@@ -44,5 +44,13 @@ export default ModelBase.extend({
 
   itemTotal: computed('price', 'quantity', function() {
     return this.price * this.quantity;
+  }),
+
+  taxIncurred: computed('event.tax', function() {
```
Comment (pradeepgangwar) [inline app/models/ticket.js]: This name is confusing. It looks like we are calculating tax incurring on each ticket but this is amount after adding tax. Something like `taxedAmount` or `ticketPriceWithTax` ot `finalPrice` would be good IMO.

File: app/controllers/public/index.js (hunk 3 above)
Comment (pradeepgangwar) [reply app/controllers/public/index.js]: see this https://github.com/fossasia/open-event-frontend/pull/2916#issuecomment-492743708

File: app/components/public/ticket-list.js (hunk 7 above)
Comment (uds5501) [reply app/components/public/ticket-list.js]: @CosmicCoder96 Please see this : https://github.com/fossasia/open-event-frontend/pull/3067#discussion_r290427205

File: app/components/public/ticket-list.js (hunk 4 above)
Comment (uds5501) [reply app/components/public/ticket-list.js]: @kushthedude @pradeepgangwar accidentally left it while testing. will remove it

File: app/templates/public/index.hbs (hunk 5 above)
Comment (uds5501) [reply app/templates/public/index.hbs]: @kushthedude sure, will do it in the next commit

File: app/models/ticket.js (hunk 9 above)
Comment (uds5501) [reply app/models/ticket.js]: @pradeepgangwar `ticketPriceWithTax` it is then :smile:

File: app/models/ticket.js (hunk 8 above)
Comment (uds5501) [reply app/models/ticket.js]: @pradeepgangwar actually `tax` is a promise object here, that's why I used `event.get('tax.isTaxIncludedInPrice')`, to resolve the promise

File: app/controllers/public/index.js (hunk 3 above)
Comment (pradeepgangwar) [reply app/controllers/public/index.js]: This looks good since now `total` on the ticker list itself contains the taxed amount.

**Comment by uds5501** [comment]
@kushthedude @shreyanshdwivedi @mrsaicharan1 @pradeepgangwar @CosmicCoder96 @niranjan94 
Ready for review.

**Comment by uds5501** [comment]
For testing purpose, you can go to this event:
https://deploy-preview-3078--open-event-frontend.netlify.com/e/178a07bc
discount code : abc(50 bucks off)

File: app/components/public/ticket-list.js [hunk 10]
```diff
@@ -94,11 +106,20 @@ export default Component.extend(FormMixin, {
...
             if (discountType === 'amount') {
               ticket.set('discount', Math.min(ticketPrice, discountValue));
+              if (taxRate) {
+                let ticketPriceWithTax = (ticketPrice - ticket.discount) * (1 + taxRate / 100);
+                ticket.set('ticketPriceWithTax', ticketPriceWithTax);
+              }
               this.discountedTickets.addObject(ticket);
             } else {
               ticket.set('discount', ticketPrice * (discountValue / 100));
+              if (taxRate) {
```
Comment (pradeepgangwar) [inline app/components/public/ticket-list.js]: This code segment is repetitive. Change it to this:
```js
let discount = discountType === 'amount' ? Math.min(ticketPrice, discountValue) : ticketPrice * (discountValue / 100);
ticket.set('discount', discount);
if (taxRate) {
  let ticketPriceWithTax = (ticketPrice - ticket.discount) * (1 + taxRate / 100);
  ticket.set('ticketPriceWithTax', ticketPriceWithTax);
}
this.discountedTickets.addObject(ticket);
```

Comment (pradeepgangwar) [reply app/components/public/ticket-list.js]: Merge this code segment with the upper one (which it repeats)

Comment (uds5501) [reply app/components/public/ticket-list.js]: @pradeepgangwar Sure! Will do that in next few minutes :smile:

**Comment by uds5501** [comment]
@pradeepgangwar @CosmicCoder96 @niranjan94 please have a look at this now.

File: app/components/public/ticket-list.js [hunk 11]
```diff
@@ -29,14 +29,20 @@ export default Component.extend(FormMixin, {
...
-      ticket => (ticket.getWithDefault('price', 0) - ticket.getWithDefault('discount', 0)) * ticket.getWithDefault('orderQuantity', 0)
-    );
+    if (this.taxInfo.get('name') !== undefined) {
+      return sumBy(this.tickets.toArray(),
+        ticket => (ticket.ticketPriceWithTax || 0) * (ticket.orderQuantity || 0)
+      );
+    } else {
+      return sumBy(this.tickets.toArray(),
+        ticket => ((ticket.price || 0) - (ticket.discount || 0)) * (ticket.orderQuantity || 0)
+      );
```
Comment (Anupam-dagar) [inline app/components/public/ticket-list.js]: Else block here isn't needed, if `if` passes then it will `return` that means the code will just execute forward in case if doesn't pass, simply writing `return` instead of else should do.
```javascript
if (this.taxInfo.get('name') !== undefined) {
      return sumBy(this.tickets.toArray(),
        ticket => (ticket.ticketPriceWithTax || 0) * (ticket.orderQuantity || 0)
      );
    } 
    return sumBy(this.tickets.toArray(),
      ticket => ((ticket.price || 0) - (ticket.discount || 0)) * (ticket.orderQuantity || 0)
      );
```

**Comment by uds5501** [comment]
@CosmicCoder96 @niranjan94 please review

**Comment by uds5501** [comment]
@CosmicCoder96 @niranjan94 @mrsaicharan1 @shreyanshdwivedi @kushthedude  please review

File: app/components/public/ticket-list.js [hunk 12]
```diff
@@ -19,6 +19,10 @@ export default Component.extend(FormMixin, {
     return !this.hasTicketsInOrder;
   }),
 
+  showTaxIncludedMessage: computed('taxInfo', function() {
+    return (this.taxInfo.get('name') && (this.taxInfo.get('rate') === 0 || this.taxInfo.get('isTaxIncludedInPrice')));
```
[the remark this answers is not in the document]
Comment (uds5501) [reply app/components/public/ticket-list.js]: @niranjan94 I was using this `.get()` because taxInfo is a promise.

[the remark this answers is not in the document]
Comment (niranjan94) [reply app/components/public/ticket-list.js]: If taxInfo is a promise, then ensure it is resolved properly _before_ use

[the remark this answers is not in the document]
Comment (niranjan94) [reply app/components/public/ticket-list.js]: If taxInfo is a promise, then ensure it is resolved properly _before_ use

[the remark this answers is not in the document]
Comment (uds5501) [reply app/components/public/ticket-list.js]: @niranjan94 Completed the required changes. Please take a look?

File: app/models/ticket.js (hunk 8 above)
Comment (CosmicCoder96) [reply app/models/ticket.js]: @uds5501 If you are explicitly calling for tax in the model, why are you using event.tax, just use tax? Since you already fetched it there?

File: app/models/ticket.js [hunk 13]
```diff
@@ -44,5 +44,13 @@ export default ModelBase.extend({
 
   itemTotal: computed('price', 'quantity', function() {
     return this.price * this.quantity;
+  }),
+
+  ticketPriceWithTax: computed('event.tax', function() {
+    let taxType = this.event.get('tax.isTaxIncludedInPrice');
+    if (!taxType && taxType !== undefined) {
```
Comment (CosmicCoder96) [inline app/models/ticket.js]: Can you elaborate on this logical comparison?

File: app/models/ticket.js [hunk 14]
```diff
@@ -44,5 +44,13 @@ export default ModelBase.extend({
...
+      return (1 + this.event.get('tax.rate') / 100) * this.price;
```
Comment (CosmicCoder96) [inline app/models/ticket.js]: use toFixed() to limit decimals.

Comment (uds5501) [reply app/models/ticket.js]: Done

File: app/models/ticket.js (hunk 13 above)
Comment (uds5501) [reply app/models/ticket.js]: I have rewritten it to use only `taxType !== undefined` just in case if there is no tax object in connection to the event.

Also, I am now using `computed('event'....)` . I have to use `event.get('tax')` because `event` is in relationship with the tickets and to access their attributes, I have to resolve the promise in a model. @CosmicCoder96

File: app/models/ticket.js (hunk 8 above)
Comment (niranjan94) [reply app/models/ticket.js]: @uds5501 I agree with @CosmicCoder96 . Resolve the tax promise at the model level and use directly in code here.

**Comment by uds5501** [comment]
@mariobehling I have updated the message for included tax price. (see in the updated screenshot)

As for the promise resolving problem, I am still working on it

File: app/models/ticket.js [hunk 15]
```diff
@@ -44,5 +44,31 @@ export default ModelBase.extend({
 
   itemTotal: computed('price', 'quantity', function() {
     return this.price * this.quantity;
+  }),
+
+  /**
+   * This attribute computes total ticket price payable after inclusion
+   * of additional taxes on the base ticket price
+   */
+  ticketPriceWithTax: computed('event', function() {
```
Comment (niranjan94) [inline app/models/ticket.js]: If you're using `tax.isTaxIncludedInPrice`, then that's what you should be watching here.

File: app/models/ticket.js [hunk 16]
```diff
@@ -44,5 +44,31 @@ export default ModelBase.extend({
...
+    }
+    return this.price;
+  }),
+
+  /**
+   * This attribute computes value added tax amount in the cases
+   * when tax amount is included in ticket price, otherwise return
+   * 0
+   */
+  includedTaxAmount: computed('event', function() {
```
Comment (niranjan94) [inline app/models/ticket.js]: Please be specific in computed properties. Watch for the exact property you are using. In this case, `event.tax.isTaxIncludedInPrice` and `event.tax.rate`

Comment (uds5501) [reply app/models/ticket.js]: apologies for repeating this mistake. Will take care from next time.

**Comment by uds5501** [comment]
@mariobehling thank you, I will resolve the conflicts and then others can review it too!

File: app/components/public/ticket-list.js [hunk 17]
```diff
@@ -19,6 +19,13 @@ export default Component.extend(FormMixin, {
     return !this.hasTicketsInOrder;
   }),
 
+  showTaxIncludedMessage: computed('taxInfo', function() {
```
[the remark this answers is not in the document]
Comment (uds5501) [reply app/components/public/ticket-list.js]: @niranjan94 sincere apologies, I missed this one in a hurry. I have covered these cases in the PR now.
